### DISCUSSION TITLE
Monsters outside submarine do not have proper pathing finding

### DISCUSSION THREAD

**Discussion by A2ur3** [FakeFishGames/Barotrauma#15298] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

They can't find path to targets and keep bumping level walls like birds bumping windows
Example 1: Crawler Broodmother keeps trying to attack human inside cave
Example 2: Hammerheads keep trying to attack submarine another side
Example 3: Latcher keeps trying to go back to abyss

### Reproduction steps

.

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Happens in both single player and multiplayer

### Version

v1.6.19.1 (Unto the Breach Update Hotfix 2)

### Which operating system did you encounter this bug on?

Windows

**Comment by A2ur3** [comment]
![20241204210119_1](https://github.com/user-attachments/assets/344527a2-d3c8-4f68-80c4-3c28bc75dbd6)
![20241204210453_1](https://github.com/user-attachments/assets/fe396dac-c227-4aae-affa-a4e1bc28f70f)
![20241204210217_1](https://github.com/user-attachments/assets/c934ae9d-5b31-4afd-910b-70202d75f8a9)

**Comment by Ek-Videogames** [comment]
Yeah this is a longstanding issue actually, but this issue is normally mitigated by the game spawning enemies only in direct line of sight of the submarine. 

However if the submarine manages to evade enemies and their AI loses track of them, or if you maneuver and use the submarine to get a terrain feature between you and the enemy then yeah enemies just blindly swim into obstructions. The map's "AI nodes" seem primarily used for the SUBMARINE pathfinding not enemy pathfinding, although the exception being the nodes inside caves.

**Comment by Regalis11** [comment]
This is actually somewhat intentional: the waypoint graph isn't dense or "complete" enough outside the submarine for monsters to reliably use it for navigation. While we could potentially generate a more complete waypoint graph outside too, it could easily lead to performance problems, and I feel it could be a little overkill for this purpose (considering you rarely see them getting stuck like this during normal gameplay).

The monsters should use certain [steering behaviors](https://code.tutsplus.com/series/understanding-steering-behaviors--gamedev-12732) to keep them from bumping against the walls like this though - I think in earlier versions they have been able to steer away from the obstacle which often allowed them to navigate around these sorts of obstacles (but not always - if they decided to go around a wall in the "wrong direction", they might not make it to the other side). It seems this has broken at some point along the way: I think they now steer so "heavily" towards the target that the obstacle avoidance behavior can't keep them from bumping against the wall.
